feat: add `math/base/special/fast/minf`

nakul-krishnakumar opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: missing_dependencies
  - task: lint_c_examples status: missing_dependencies
  - task: lint_c_benchmarks status: missing_dependencies
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds `math/base/special/fast/minf` .

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

nakul-krishnakumar committed:
feat: add `math/base/special/fast/minf`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
If it’s not currently being reviewed, I can take a look at it! 😄

kgryte commented:
@anandkaranubc Go for it!

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/fast/minf/lib/main.js:
```diff
@@ -0,0 +1,67 @@
+* @example
+* var v = minf( NaN, 3.14 );
+* // returns ~3.14
+*
```
Is this behaviour expected? @kgryte

I see the same in `special/fast/min` but not in `special/min`

anandkaranubc replied:
Sorry my bad. Just noticed the this line:

`* -   The function ignores the sign of `0` and does not check for `NaN` arguments.`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/fast/minf/lib/main.js:
```diff
@@ -0,0 +1,67 @@
+* var v = minf( NaN, 3.14 );
+* // returns ~3.14
+*
+* @example
+* var v = minf( -0.0, +0.0 );
+* // returns +0.0
```
Same question for this...

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/fast/minf/src/main.c:
```diff
@@ -0,0 +1,41 @@
+* @param y       second number
+* @return        minimum value
+*
+* @example
+* float v = stdlib_base_fast_minf( 3.14f, 4.2f );
+* // returns 3.14f
+*
+* @example
+* float v = stdlib_base_fast_minf( 0.0f, -0.0f );
+* // returns -0.0f
```
```suggestion
* @example
* float v = stdlib_base_fast_minf( 3.14f, 4.2f );
* // returns 3.14f
```

Comparing to other implementations, we usually put only one example for `main.c`

kgryte replied:
This is fine. No need to change.

kgryte replied:
For context, there isn't an overarching reason to limit to one example. If and when we have doctesting for C source files, having multiple examples would actually be a good thing, as the examples can be used as a lightweight form of unit testing. Historically, we've only added one example, if any, due to only displaying a usage example and minimizing maintenance effort. If we had C doctesting, that would change the equation a bit more.

anandkaranubc replied:
Makes sense. Thanks for explaining!

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/fast/minf/examples/index.js:
```diff
@@ -0,0 +1,34 @@
+* limitations under the License.
+*/
+
+'use strict';
+
+var minstd = require( '@stdlib/random/base/minstd-shuffle' );
+var minf = require( './../lib' );
+
+var x;
+var y;
+var v;
+var i;
+
+for ( i = 0; i < 100; i++ ) {
+	x = minstd();
+	y = minstd();
+	v = minf( x, y );
+	console.log( 'minf(%d,%d) = %d', x, y, v );
+}
```
@nakul-krishnakumar Do you mind changing this to use `logEachMap` instead, as done in `fast/maxf`, for consistency?

nakul-krishnakumar replied:
Sure!

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/fast/minf/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,138 @@
+	struct timeval now;
+	gettimeofday( &now, NULL );
+	return (double)now.tv_sec + (double)now.tv_usec/1.0e6;
+}
+
+/**
+* Generates a random number on the interval [0,1).
+*
+* @return random number
+*/
+static float rand_float( void ) {
+	int r = rand();
+	return (float)r / ( (float)RAND_MAX + 1.0f );
+}
```
```suggestion
/**
* Generates a random number on the interval [min,max).
*
* @param min    minimum value (inclusive)
* @param max    maximum value (exclusive)
* @return       random number
*/
static float random_uniform( const float min, const float max ) {
	float v = (float)rand() / ( (float)RAND_MAX + 1.0f );
	return min + ( v*(max-min) );
}
```

For consistency too, according to the latest project conventions. You can check `math/base/special/sinf` for reference.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/fast/minf/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,138 @@
+	float z;
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = ( 200.0f * rand_float() ) - 100.0f;
+		y[ i ] = ( 200.0f * rand_float() ) - 100.0f;
+	}
```
These will be updated accordingly.

nakul-krishnakumar committed:
chore: switch to `logEachMap` according to code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/fast/minf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,67 @@
+
+// MAIN //
+
+bench(pkg + '::native', opts, function benchmark( b ) {
```
```suggestion
bench( pkg+'::native', opts, function benchmark( b ) {
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/fast/minf/README.md:
````diff
@@ -0,0 +1,234 @@
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var minstd = require( '@stdlib/random/base/minstd-shuffle' );
+var minf = require( '@stdlib/math/base/special/fast/minf' );
+
+var x;
+var y;
+var v;
+var i;
+
+for ( i = 0; i < 100; i++ ) {
+    x = minstd();
+    y = minstd();
+    v = minf( x, y );
+    console.log( 'minf(%d,%d) = %d', x, y, v );
+}
````
This will be updated too.

anandkaranubc reviewed (changes requested):
Rest looks good! Thank you!

nakul-krishnakumar committed:
chore: update according to code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

nakul-krishnakumar committed:
style: update according to code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

nakul-krishnakumar committed:
docs: update according to code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

nakul-krishnakumar commented:
I have made the necessary changes, thank you for the review! @anandkaranubc

anandkaranubc reviewed (approved):
Everything looks good! Thanks for making the changes @nakul-krishnakumar! 🚀

kgryte reviewed (approved):
LGTM
